### PR TITLE
Fix Coverage Reports on Jenkins

### PR DESCRIPTION
**Pull request by ravi-mosaicml**
Jenkins Coverage reports were broken -- the coverage percentage would vary wildly between (passing) runs, indicating a potential issue with merging coverage reports and displaying them under one file. This PR fixes that by archiving the raw coverage files, and manually merging them after all tests are run, instead of relying on Jenkin's merge coverage report option.

Closes https://mosaicml.atlassian.net/browse/CO-772

### CODE REVIEW

**Comment by ravi-mosaicml** [comment]
Closing this PR as stale.

**Review by mvpatel2000 — APPROVED** [review]
Approving, but I'll be honest -- I don't really understand Jenkins so I'm not super sure what's going on here. I guess we'll discuss Monday, but trusting you tested this enough to merge.

**Comment by mvpatel2000** [comment]
Hm... looks like I'm not codeowner so Bandish will have to approve

File: .ci/Jenkinsfile
Comment (bandish-shah) [inline .ci/Jenkinsfile:450]: [nit] Not a huge fan of embedding YAML's in functional code like this.  Could we make this a config YAML job template somewhere, read it in and populate the required config fields before passing it to `podArgs`?

File: .ci/Jenkinsfile
Comment (bandish-shah) [inline .ci/Jenkinsfile:503]: [nit] Similar comment to above, not a huge fan of embedding bash scripts as strings in groovy.  Would be nice if we just converted this to a util script or something and called it vs. embedding this in functional code.

**Review by bandish-shah — APPROVED** [review]
LGTM, approving to unblock but see nits.

File: .ci/Jenkinsfile
Comment (ravi-mosaicml) [reply .ci/Jenkinsfile]: Good point; done.

File: .ci/Jenkinsfile
Comment (ravi-mosaicml) [reply .ci/Jenkinsfile:450]: Unfortunately, for security reasons, we can't easily. If the yaml is in a separate file, then external contributors could modify the yaml to create arbitrary pods on our cluster. Changes to the Jenkinsfile itself are protected, though -- Jenkins won't build PRs that modify the Jenkinsfile unless the commit author is from Mosaic (validated via the GH api, I presume).

File: .ci/Jenkinsfile
Comment (ravirahman) [inline .ci/Jenkinsfile:503]:
```suggestion
                        sh "./.ci/merge_coverage_reports.sh $buildOutputFolder $mergedCoverageFile"
```

**Pull request merged**
